src/backend/metal/src/command.rs

```diff
@@ -4228,26 +4311,50 @@ impl com::RawCommandBuffer<Backend> for CommandBuffer {
             .issue_many(commands);
     }
 
-    unsafe fn set_event(&mut self, _: &(), _: pso::PipelineStage) {
-        unimplemented!()
+    unsafe fn set_event(&mut self, event: &native::Event, _: pso::PipelineStage) {
+        self.inner
+            .borrow_mut()
+            .events
+            .push((Arc::clone(&event.0), true));
     }
 
-    unsafe fn reset_event(&mut self, _: &(), _: pso::PipelineStage) {
-        unimplemented!()
+    unsafe fn reset_event(&mut self, event: &native::Event, _: pso::PipelineStage) {
+        self.inner
+            .borrow_mut()
+            .events
+            .push((Arc::clone(&event.0), false));
     }
 
     unsafe fn wait_events<'a, I, J>(
         &mut self,
-        _: I,
-        _: Range<pso::PipelineStage>,
-        _: J
+        events: I,
+        stages: Range<pso::PipelineStage>,
+        barriers: J
     ) where
         I: IntoIterator,
-    I::Item: Borrow<()>,
-    J: IntoIterator,
-    J::Item: Borrow<memory::Barrier<'a, Backend>>,
+        I::Item: Borrow<native::Event>,
+        J: IntoIterator,
+        J::Item: Borrow<memory::Barrier<'a, Backend>>,
     {
-        unimplemented!()
+        let mut need_barrier = false;
+
+        for event in events {
+            let mut inner = self.inner.borrow_mut();
+            let event = &event.borrow().0;
+            let is_local = inner.events
+                .iter()
+                .rfind(|ev| Arc::ptr_eq(&ev.0, event))
+                .map_or(false, |ev| ev.1);
+            if is_local {
+                need_barrier = true;
+            } else {
+                inner.host_events.push(Arc::clone(event));
+            }
+        }
+
+        if need_barrier {
+            self.pipeline_barrier(stages, memory::Dependencies::empty(), barriers);
```

Comment:
I think events could allow the pipeline barrier operation to be delayed until the data is actually used, so it could be non-optimal to force the pipeline barrier at this point.

Although none of this really matters as long as we continue to no-op `pipeline_barrier`